File: src/seni/Interp.js
/*
 *  Seni
 *  Copyright (C) 2015  Inderjit Gill <email@indy.io>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import PublicBinding from './PublicBinding';
import Immutable from 'immutable';
import MathUtil from './MathUtil';


// the following map* functions work in the range 0..1:

function mapLinear(x) {
  return x;
}

function mapQuickEase(x) {
  const x2 = x * x;
  const x3 = x * x * x;
  return (3 * x2) - (2 * x3);
}

function mapSlowEaseIn(x) {
  const s = Math.sin(x * MathUtil.PIbyTwo);
  return s * s * s * s;
}

function mapSlowEaseInEaseOut(x) {
  return x - (Math.sin(x * MathUtil.TAU) / MathUtil.TAU);
}

const remappingFn = new Immutable.Map({'linear': mapLinear,
                                       'quick': mapQuickEase,
                                       'slow-in': mapSlowEaseIn,
                                       'slow-in-out': mapSlowEaseInEaseOut});

function remapFn(params) {

  const from = params.from || [0, 1];
  const to = params.to || [0, 100];
  const clamping = params.clamping || false;
  const mapping = params.mapping || 'linear';

  const [fromA, fromB] = from;
  const [toA, toB] = to;
  const [fromM, fromC] = MathUtil.mc([fromA, 0], [fromB, 1]);
  const [toM, toC] = MathUtil.mc([0, toA], [1, toB]);

  let normalisedMappingFn = remappingFn.get(mapping);

  if (normalisedMappingFn === undefined) {
    normalisedMappingFn = remappingFn.get('linear');
  }

  return function(parameters) {
    const val = parameters.val || 0;
    const fromInterp = (fromM * val) + fromC;
    const toInterp = normalisedMappingFn(fromInterp);
    const res = (toM * toInterp) + toC;

    if (clamping) {
      return fromInterp < 0 ? toA : (fromInterp > 1) ? toB : res;
    } else {
      return res;
    }
  };
}

function interpolate(a, b, t) {
  return (a * (1 - t)) + (b * t);
}


function makeBezierFn(coords) {
  return function(parameters) {
    const t = parameters.t || 0;
    let x = MathUtil.bezierPoint(coords[0][0],
                                 coords[1][0],
                                 coords[2][0],
                                 coords[3][0],
                                 t);
    let y = MathUtil.bezierPoint(coords[0][1],
                                 coords[1][1],
                                 coords[2][1],
                                 coords[3][1],
                                 t);
    return [x, y];
  };
}

function makeBezierTangentFn(coords) {
  return function(parameters) {
    const t = parameters.t || 0;
    let x = MathUtil.bezierTangent(coords[0][0],
                                   coords[1][0],
                                   coords[2][0],
                                   coords[3][0],
                                   t);
    let y = MathUtil.bezierTangent(coords[0][1],
                                   coords[1][1],
                                   coords[2][1],
                                   coords[3][1],
                                   t);
    return [x, y];
  };
}

const defaultCoords = [[440, 400],
                       [533, 700],
                       [766, 200],
                       [900, 500]];

const Interp = {
  publicBindings: [
    new PublicBinding(
      'interp/fn',
      ``,
      {},
      () => remapFn
    ),

    new PublicBinding(
      'interp/bezier',
      ``,
      {coords: defaultCoords, t: 1},
      (self) => function(params) {
        const {coords, t} = self.mergeWithDefaults(params);
        // make a Bezier fn and then invoke it straight away with the t value
        return makeBezierFn(coords)({t: t});
      }
    ),

    new PublicBinding(
      'interp/bezier-fn',
      ``,
      {coords: defaultCoords},
      (self) => function(params) {
        const {coords} = self.mergeWithDefaults(params);
        // return a function that only accepts a t value
        return makeBezierFn(coords);
      }
    ),

    new PublicBinding(
      'interp/bezier-tangent',
      ``,
      {coords: defaultCoords, t: 1},
      (self) => function(params) {
        const {coords, t} = self.mergeWithDefaults(params);
        return makeBezierTangentFn(coords)({t: t});
      }
    ),

    new PublicBinding(
      'interp/bezier-tangent-fn',
      ``,
      {coords: defaultCoords},
      (self) => function(params) {
        const {coords} = self.mergeWithDefaults(params);

        // return a function that only accepts a t value
        return makeBezierTangentFn(coords);
      }
    ),

    new PublicBinding(
      'interp/circle',
      ``,
      {position: [0, 0],
       radius: 1,
       t: 0
      },
      (self) => function(params) {
        const {position, radius, t} = self.mergeWithDefaults(params);

        let [x, y] = position;

        let angle = t * MathUtil.TAU;

        let vx = (Math.sin(angle) * radius) + x;
        let vy = (Math.cos(angle) * radius) + y;

        return [vx, vy];
      }
    )
  ],
  interpolate,
  remapFn
};

export default Interp;

File: core/sen-core/src/interp.rs
// Copyright (C) 2019 Inderjit Gill

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

use crate::ease::*;
use crate::mathutil::*;
use crate::vm::InterpStateStruct;

pub fn interp_from_struct(t: f32, interp_state: &InterpStateStruct) -> f32 {
    let from_interp = (interp_state.from_m * t) + interp_state.from_c;
    let to_interp = easing(from_interp, interp_state.mapping);
    let res = (interp_state.to_m * to_interp) + interp_state.to_c;

    if interp_state.clamping {
        return if from_interp < 0.0 {
            interp_state.to.0
        } else if from_interp > 1.0 {
            interp_state.to.1
        } else {
            res
        };
    }

    res
}

pub fn interp_scalar(a: f32, b: f32, mapping: Easing, clamping: bool, t: f32) -> f32 {
    let new_t = easing(t, mapping);
    let res = lerp(new_t, a, b);

    if clamping {
        if new_t < 0.0 {
            return a;
        } else if new_t > 1.0 {
            return b;
        } else {
            return res;
        }
    }

    return res;
}

pub fn interp_cos(amplitude: f32, frequency: f32, t: f32) -> f32 {
    amplitude * (t * frequency).cos()
}

pub fn interp_sin(amplitude: f32, frequency: f32, t: f32) -> f32 {
    amplitude * (t * frequency).sin()
}

pub fn interp_bezier(coords: &[f32; 8], t: f32) -> (f32, f32) {
    (
        bezier_point(coords[0], coords[2], coords[4], coords[6], t),
        bezier_point(coords[1], coords[3], coords[5], coords[7], t),
    )
}

pub fn interp_bezier_tangent(coords: &[f32; 8], t: f32) -> (f32, f32) {
    (
        bezier_tangent(coords[0], coords[2], coords[4], coords[6], t),
        bezier_tangent(coords[1], coords[3], coords[5], coords[7], t),
    )
}

pub fn interp_circle(position: (f32, f32), radius: f32, t: f32) -> (f32, f32) {
    let angle = t * TAU;

    (
        (angle.sin() * radius) + position.0,
        (angle.cos() * radius) + position.1,
    )
}

pub fn interp_ray(point: (f32, f32), direction: (f32, f32), t: f32) -> (f32, f32) {
    // direction should be a normalized vector
    (point.0 + (direction.0 * t), point.1 + (direction.1 * t))
}

#[cfg(test)]
mod tests {
    use crate::vm::tests::*;

    #[test]
    fn test_interp_cos() {
        is_debug_str(
            "(loop (x upto: 5)
                   (probe scalar: (interp/cos t: (/ x 5))))",
            "1 0.9800666 0.921061 0.8253356 0.6967067 0.5403023",
        );
        is_debug_str(
            "(loop (x upto: 5)
                   (probe scalar: (interp/cos amplitude: 3 t: (/ x 5))))",
            "3 2.9401999 2.7631829 2.476007 2.09012 1.6209068",
        );
        is_debug_str(
            "(loop (x upto: 5)
                   (probe scalar: (interp/cos frequency: 7 t: (/ x 5))))",
            "1 0.16996716 -0.9422223 -0.49026057 0.7755658 0.75390226",
        );
    }

    #[test]
    fn test_interp_sin() {
        is_debug_str(
            "(loop (x upto: 5)
                   (probe scalar: (interp/sin t: (/ x 5))))",
            "0 0.19866933 0.38941833 0.5646425 0.7173561 0.84147096",
        );
        is_debug_str(
            "(loop (x upto: 5)
                   (probe scalar: (interp/sin amplitude: 3 t: (/ x 5))))",
            "0 0.596008 1.168255 1.6939275 2.1520681 2.5244129",
        );
        is_debug_str(
            "(loop (x upto: 5)
                   (probe scalar: (interp/sin frequency: 7 t: (/ x 5))))",
            "0 0.98544973 0.3349882 -0.8715759 -0.6312667 0.6569866",
        );
    }

    #[test]
    fn test_interp_build() {
        is_debug_str(
            "(define i (interp/build from: [0 1] to: [0 100]))
             (probe scalar: (interp/value from: i t: 0.5))",
            "50",
        );
        is_debug_str(
            "(define i (interp/build from: [10 20] to: [50 200]))
             (probe scalar: (interp/value from: i t: 10.0))",
            "50",
        );
        is_debug_str(
            "(define i (interp/build from: [10 20] to: [50 200]))
             (probe scalar: (interp/value from: i t: 20.0))",
            "200",
        );
        is_debug_str(
            "(define i (interp/build from: [50 10] to: [100 1000]))
             (probe scalar: (interp/value from: i t: 50.0))",
            "100",
        );
        is_debug_str(
            "(define i (interp/build from: [50 10] to: [100 1000]))
             (probe scalar: (interp/value from: i t: 10.0))",
            "1000",
        );

        // clamping
        is_debug_str(
            "(define i (interp/build from: [0 1] to: [0 100] clamping: false))
             (probe scalar: (interp/value from: i t: 2.0))",
            "200",
        );
        is_debug_str(
            "(define i (interp/build from: [0 1] to: [0 100] clamping: true))
             (probe scalar: (interp/value from: i t: 2.0))",
            "100",
        );
        is_debug_str(
            "(define i (interp/build from: [0 1] to: [0 100] clamping: true))
             (probe scalar: (interp/value from: i t: -2.0))",
            "0",
        );
    }

}
